src/epoch_credits/mod.rs

```diff
@@ -1,35 +1,176 @@
+    ) -> anyhow::Result<(Vec<StakingApy>, Vec<ValidatorReward>)> {
+        let mut staking_seen_voters = BTreeSet::new();
+        let mut staking_apys = Vec::new();
+        let mut validator_rewards = Vec::new();
+
+        for Reward {
+            pubkey,
+            lamports,
+            post_balance,
+            reward_type,
+        } in rewards
+        {
+            match reward_type {
+                Some(RewardType::Staking) => {
+                    let account_info = self
+                        .client
+                        .get_account(&Pubkey::try_from(pubkey.as_ref())?)?;
+                    let stake_state: StakeState = bincode::deserialize(&account_info.data)?;
+                    if let Some(delegation) = stake_state.delegation() {
+                        let voter = format!("{}", delegation.voter_pubkey);
+                        if !staking_seen_voters.contains(&voter) && lamports > 0 {
+                            let lamports = lamports as u64;
+                            let prev_balance = post_balance - lamports;
+                            let epoch_rate = lamports as f64 / prev_balance as f64;
+                            let apy = 100.0
+                                * (f64::powf(1.0 + epoch_rate / epochs_in_year, epochs_in_year)
+                                    - 1.0);
+                            staking_apys.push(StakingApy {
+                                voter: voter.clone(),
+                                percent: apy,
+                            });
+                            staking_seen_voters.insert(voter);
+                        }
+                    }
+                }
+                Some(RewardType::Voting) => validator_rewards.push(ValidatorReward {
+                    voter: pubkey,
+                    lamports: post_balance,
+                }),
+                _ => (), // TODO other reward types
+            }
+        }
+        Ok((staking_apys, validator_rewards))
+    }
+
+    /// Gets the rewards for `epoch` optionally given `epoch_info`. Returns `Ok(None)` if there
+    /// haven't been any rewards in the current epoch yet, `Ok(Some(rewards))` if there have, and
+    /// otherwise returns an error.
+    fn get_rewards_for_epoch(
+        &self,
+        epoch: Epoch,
+        epoch_info: Option<EpochInfo>,
+    ) -> anyhow::Result<Option<Rewards>> {
+        let info = epoch_info.unwrap_or(self.client.get_epoch_info()?);
+
+        // Convert epoch number to slot
+        let start_slot = epoch * info.slots_in_epoch;
+
+        // We cannot use an excessively large range if the epoch just started. There is a chance that
+        // the end slot has not been reached and strange behaviour will occur.
+        // If this is the current epoch and less than `SLOT_OFFSET` slots have elapsed, then do not define an
+        // end_slot for use in the RPC call.
+        let end_slot = if info.epoch == epoch && info.slot_index < SLOT_OFFSET {
```

Comment:
If the optional `EpochInfo` is for looking at historical records, then this block needs to be skipped as all historical lookups would always have enough `SLOT_OFFSET`.

Reply:
Lets require `epoch_info` to be current. This was my original intention anyway.